### ISSUE TITLE
Install defaults to master branch not git default

### ISSUE THREAD

**Issue by SyntaxRules** [npm/npm#8178]
If I want to install a package that isn't in the npm repository I would perform the following command:

```
 npm install https://github.com/SyntaxRules/not-master.git
```

Or I could put it in my package.json like so: 
```
"not-master": "https://github.com/SyntaxRules/not-master.git"
```

**But this fails** because there is no `master` branch in the `not-master` repository. Instead the default branch is `0.x`. npm should automatically pull the default branch. Instead we are assuming that master is the default branch.

I know that I could just specify with `https://github.com/SyntaxRules/not-master.git#0.x`, but if the repository owner is going to release `1.x` as a new branch (and make that the default) then every time a major upgrade happens every user has to upgrade their package.json.

The error log:
```shell
0 info it worked if it ends with ok
1 verbose cli [ 'c:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'c:\\Users\\someUser\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install',
1 verbose cli   'https://github.com/SyntaxRules/not-master.git' ]
2 info using npm@2.7.4
3 info using node@v0.12.2
4 silly cache add args [ 'https://github.com/SyntaxRules/not-master.git', null ]
5 verbose cache add spec https://github.com/SyntaxRules/not-master.git
6 silly cache add parsed spec { raw: 'https://github.com/SyntaxRules/not-master.git',
6 silly cache add   scope: null,
6 silly cache add   name: null,
6 silly cache add   rawSpec: 'https://github.com/SyntaxRules/not-master.git',
6 silly cache add   spec: 'git+ssh://git@github.com/SyntaxRules/not-master.git',
6 silly cache add   type: 'hosted',
6 silly cache add   hosted:
6 silly cache add    { type: 'github',
6 silly cache add      ssh: 'git@github.com:SyntaxRules/not-master.git',
6 silly cache add      sshUrl: 'git+ssh://git@github.com/SyntaxRules/not-master.git',
6 silly cache add      httpsUrl: 'https://github.com/SyntaxRules/not-master.git',
6 silly cache add      directUrl: 'https://raw.githubusercontent.com/SyntaxRules/not-master/master/package.json' } }
7 info maybeGithub Attempting https://github.com/SyntaxRules/not-master.git from git://github.com/SyntaxRules/not-master.git
8 verbose addRemoteGit git-github-com-SyntaxRules-not-master-git-2497b0b7 not in flight; caching
9 info git [ 'config', '--get', 'remote.origin.url' ]
10 verbose addRemoteGit remote.origin.url: git://github.com/SyntaxRules/not-master.git
11 verbose addRemoteGit updating existing cached remote C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-2497b0b7
12 info git [ 'fetch', '-a', 'origin' ]
13 verbose addRemoteGit git fetch -a origin (git://github.com/SyntaxRules/not-master.git)
14 verbose addRemoteGit skipping chownr on Windows
15 verbose addRemoteGit original treeish: master
16 info git [ 'rev-list', '-n1', 'master' ]
17 error git -c core.longpaths=true rev-list -n1 master
18 error git rev-list -n1 master: fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
18 error git rev-list -n1 master: Use '--' to separate paths from revisions, like this:
18 error git rev-list -n1 master: 'git <command> [<revision>...] -- [<file>...]'
19 info maybeGithub Couldn't clone git://github.com/SyntaxRules/not-master.git
20 info maybeGithub Now attempting https://github.com/SyntaxRules/not-master.git from git@github.com:SyntaxRules/not-master.git
21 verbose addRemoteGit git-github-com-SyntaxRules-not-master-git-81840964 not in flight; caching
22 info git [ 'config', '--get', 'remote.origin.url' ]
23 error git -c core.longpaths=true config --get remote.origin.url
24 verbose addRemoteGit remote.origin.url:
25 warn addRemoteGit Error: Command failed: git -c core.longpaths=true config --get remote.origin.url
25 warn addRemoteGit
25 warn addRemoteGit     at ChildProcess.exithandler (child_process.js:751:12)
25 warn addRemoteGit     at ChildProcess.emit (events.js:110:17)
25 warn addRemoteGit     at maybeClose (child_process.js:1015:16)
25 warn addRemoteGit     at Process.ChildProcess._handle.onexit (child_process.js:1087:5)
25 warn addRemoteGit  resetting remote C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964 because of error: { [Error: Command failed: git -c core.longpaths=true config --get remote.origin.url
25 warn addRemoteGit ]
25 warn addRemoteGit   killed: false,
25 warn addRemoteGit   code: 1,
25 warn addRemoteGit   signal: null,
25 warn addRemoteGit   cmd: 'git -c core.longpaths=true config --get remote.origin.url' }
26 info addRemoteGit resetting C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964
27 silly gentlyRm C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964 is being purged
28 verbose gentlyRm don't care about contents; nuking C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964
29 silly vacuum-fs purging C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964
30 silly vacuum-fs quitting because other entries in C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes
31 info git [ 'clone',
31 info git   '--template=C:\\Users\\someUser\\AppData\\Roaming\\npm-cache\\_git-remotes\\_templates',
31 info git   '--mirror',
31 info git   'git@github.com:SyntaxRules/not-master.git',
31 info git   'C:\\Users\\someUser\\AppData\\Roaming\\npm-cache\\_git-remotes\\git-github-com-SyntaxRules-not-master-git-81840964' ]
32 error git -c core.longpaths=true clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964
33 error git clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964: Cloning into bare repository 'C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964'...
33 error git clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964: ssh: connect to host github.com port 22: Bad file number
33 error git clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964: fatal: Could not read from remote repository.
33 error git clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964:
33 error git clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964: Please make sure you have the correct access rights
33 error git clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964: and the repository exists.
34 verbose stack Error: Command failed: git -c core.longpaths=true clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964
34 verbose stack Cloning into bare repository 'C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964'...
34 verbose stack ssh: connect to host github.com port 22: Bad file number
34 verbose stack fatal: Could not read from remote repository.
34 verbose stack
34 verbose stack Please make sure you have the correct access rights
34 verbose stack and the repository exists.
34 verbose stack
34 verbose stack     at ChildProcess.exithandler (child_process.js:751:12)
34 verbose stack     at ChildProcess.emit (events.js:110:17)
34 verbose stack     at maybeClose (child_process.js:1015:16)
34 verbose stack     at Process.ChildProcess._handle.onexit (child_process.js:1087:5)
35 verbose cwd c:\projects
36 error Windows_NT 6.3.9600
37 error argv "c:\\Program Files\\nodejs\\node.exe" "c:\\Users\\someUser\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install" "https://github.com/SyntaxRules/not-master.git"
38 error node v0.12.2
39 error npm  v2.7.4
40 error code 128
41 error Command failed: git -c core.longpaths=true clone --template=C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\_templates --mirror git@github.com:SyntaxRules/not-master.git C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964
41 error Cloning into bare repository 'C:\Users\someUser\AppData\Roaming\npm-cache\_git-remotes\git-github-com-SyntaxRules-not-master-git-81840964'...
41 error ssh: connect to host github.com port 22: Bad file number
41 error fatal: Could not read from remote repository.
41 error
41 error Please make sure you have the correct access rights
41 error and the repository exists.
42 error If you need help, you may report this error at:
42 error     <https://github.com/npm/npm/issues>
43 verbose exit [ 1, true ]

```

**Comment by othiym23** [comment]
There's a couple things going on in that debug log, both of which appear to be unrelated to the problem you described:

1. The actual syntax to install a dependency from Git (and when putting that dependency in `package.json`) is `npm install git+https://github.com/SyntaxRules/not-master.git`.
2. `npm@2.7.4` has a bunch of problems with its git dependency handling (which it acquired in the process of fixing other problems with git dependency handling). Most of those issues are addressed as of `npm@2.9.0` (aka `latest`), so upgrade to latest when dealing with git issues.

Even with the latest npm, trying to install that repo will fail, pretty much for the exact reason you describe: npm follows the GitHub convention of treating `master` as the default branch. @isaacs and I talked this use case through, and we're in agreement that it doesn't make sense to add a bunch of extra, reasonably frangible logic to read the repo configuration to check what branch to treat as default when it's straightforward to just specify the desired branch with `#branchname`. Note that "default branch" is in itself a GitHubism, and the actual situation is [a little more complicated](http://stackoverflow.com/questions/18726037/default-branch-after-git-clone) (or simpler, depending on how you look at it).

> I know that I could just specify with https://github.com/SyntaxRules/not-master.git#0.x, but if the repository owner is going to release 1.x as a new branch (and make that the default) then every time a major upgrade happens every user has to upgrade their package.json.

That seems appropriate from a semver-y point of view, doesn't it? It does to me. If that's a concern for you, you can always create a branch `for-use-with-npm` (or similar) that is set to point at the branch you want to track. That seems analogous to how you'd use `dist-tag`s with packages on a registry.

I hope this answers your question. We've put a bunch of time into improving our git story recently, and some of that is trying to simplify and make more consistent how the git-handling logic works. I hope it makes sense that we can't try to deal with every possible edge case of mapping a package system onto a version control system, and I hope my explanation is at least plausible, if not 100% persuasive. Thanks for taking the time to put together a careful explanation of the problem with a repro case. I appreciate it!

**Comment by SyntaxRules** [comment]
Agreed, this is not a npm issue, or bug. This is a feature-request. I appreciate all that has been done on npm's side to fix git problems (I've been following the releases and have been using multiple versions for testing). For the sake of completeness I've upgraded to 2.9.1 and used the correct npm install syntax. You'll find the log at the bottom of this comment.

As for adding this feature, I can respect your decision to add it or not add it. I'd just like to add a few arguments to the table.

> npm follows the GitHub convention of treating master as the default branch.

I disagree that this is GitHub convention, by default you get a master, and most people just never bother to change it. You can change the default branch in GitHub to be whatever you want in the repo settings (as I'm sure you already knew). Perhaps this whole issue was logged because I hate the ambiguous name of master. Some people use it as stable, others use it as development. Where do I go if I want to see the 1.x version of the code (and even build off of it)? How about 2.x?

> it doesn't make sense to add a bunch of extra, reasonably frangible logic to read the repo configuration to check what branch to treat as default

Agreed, I wouldn't want a lot of frangible logic in there either. But I hope this isn't the case. Consider the example of cloning my repo. I would perform a git clone with:

```
git clone https://github.com/SyntaxRules/not-master.git
```

I get my repo, and the branch I start out in is `0.x`. This is just given to me by git *as long as I don't tell it what branch to get*. Can we use this same idea with npm? `Line 16` in the log below makes me think that somewhere in npm we decided to pull master because a branch wasn't specified. If you don't specify a branch to pull, then you'll just get the default. Right?

Could we just say, if the user hasn't specified a branch, just use the default that git pulls down. (Which isn't necessarily master.)

> That seems appropriate from a semver-y point of view, doesn't it? It does to me.

Ok, but I want to avoid making people update their code to the next version if they already know the want to live life on the bleeding edge. And you are right, I could do this with another branch, but then when I take in a pull request for the latest version, I have to copy it over to this other branch as well. Someone will forget this. I'm either lazy or paranoid- take your pick. If you decide not to do anything about this feature, this is probably the direction I'll go. (I wont call it master though :) )

Thanks!

The log:
```shell
0 info it worked if it ends with ok
1 verbose cli [ 'c:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'c:\\Users\\someUser\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install',
1 verbose cli   'git+https://github.com/SyntaxRules/not-master.git' ]
2 info using npm@2.9.1
3 info using node@v0.12.2
4 verbose install initial load of c:\projects\package.json
5 verbose readDependencies loading dependencies from c:\projects\package.json
6 silly cache add args [ 'git+https://github.com/SyntaxRules/not-master.git', null ]
7 verbose cache add spec git+https://github.com/SyntaxRules/not-master.git
8 silly cache add parsed spec { raw: 'git+https://github.com/SyntaxRules/not-master.git',
8 silly cache add   scope: null,
8 silly cache add   name: null,
8 silly cache add   rawSpec: 'git+https://github.com/SyntaxRules/not-master.git',
8 silly cache add   spec: 'git+https://github.com/SyntaxRules/not-master.git',
8 silly cache add   type: 'hosted',
8 silly cache add   hosted:
8 silly cache add    { type: 'github',
8 silly cache add      ssh: 'git@github.com:SyntaxRules/not-master.git',
8 silly cache add      sshUrl: 'git+ssh://git@github.com/SyntaxRules/not-master.git',
8 silly cache add      httpsUrl: 'git+https://github.com/SyntaxRules/not-master.git',
8 silly cache add      gitUrl: 'git://github.com/SyntaxRules/not-master.git',
8 silly cache add      shortcut: 'github:SyntaxRules/not-master',
8 silly cache add      directUrl: 'https://raw.githubusercontent.com/SyntaxRules/not-master/master/package.json' } }
9 verbose addRemoteGit caching git+https://github.com/SyntaxRules/not-master.git
10 verbose addRemoteGit git+https://github.com/SyntaxRules/not-master.git is a repository hosted by github
11 silly tryClone cloning git+https://github.com/SyntaxRules/not-master.git via git+https://github.com/SyntaxRules/not-master.git
12 verbose tryClone https-github-com-SyntaxRules-not-master-git-555e85a1 not in flight; caching
13 info git [ 'clone',
13 info git   '--template=C:\\Users\\someUser\\AppData\\Roaming\\npm-cache\\_git-remotes\\_templates',
13 info git   '--mirror',
13 info git   'https://github.com/SyntaxRules/not-master.git',
13 info git   'C:\\Users\\someUser\\AppData\\Roaming\\npm-cache\\_git-remotes\\https-github-com-SyntaxRules-not-master-git-555e85a1' ]
14 verbose mirrorRemote git+https://github.com/SyntaxRules/not-master.git git clone https://github.com/SyntaxRules/not-master.git
15 verbose setPermissions git+https://github.com/SyntaxRules/not-master.git skipping chownr on Windows
16 verbose resolveHead git+https://github.com/SyntaxRules/not-master.git original treeish: master
17 info git [ 'rev-list', '-n1', 'master' ]
18 error git rev-list -n1 master: fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
18 error git rev-list -n1 master: Use '--' to separate paths from revisions, like this:
18 error git rev-list -n1 master: 'git <command> [<revision>...] -- [<file>...]'
19 verbose stack Error: Command failed: git -c core.longpaths=true rev-list -n1 master
19 verbose stack fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
19 verbose stack Use '--' to separate paths from revisions, like this:
19 verbose stack 'git <command> [<revision>...] -- [<file>...]'
19 verbose stack
19 verbose stack     at ChildProcess.exithandler (child_process.js:751:12)
19 verbose stack     at ChildProcess.emit (events.js:110:17)
19 verbose stack     at maybeClose (child_process.js:1015:16)
19 verbose stack     at Process.ChildProcess._handle.onexit (child_process.js:1087:5)
20 verbose cwd c:\projects
21 error Windows_NT 6.3.9600
22 error argv "c:\\Program Files\\nodejs\\node.exe" "c:\\Users\\someUser\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install" "git+https://github.com/SyntaxRules/not-master.git"
23 error node v0.12.2
24 error npm  v2.9.1
25 error code 128
26 error Command failed: git -c core.longpaths=true rev-list -n1 master
26 error fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
26 error Use '--' to separate paths from revisions, like this:
26 error 'git <command> [<revision>...] -- [<file>...]'
27 error If you need help, you may report this error at:
27 error     <https://github.com/npm/npm/issues>
28 verbose exit [ 1, true ]
```
